Tof erf mapping

robbietuk opened the pull request:
Development PR for computing and storing erf values in memory. @KrisThielemans @ALEXJAZZ008008 
________

TODO
----

- [x] Implement BSpline
- [x] ~~Use symetry `erf(x) = -erf(-x)`~~
- [x] Implement out boundary for BSplines
- [x] Implement linear and nearest neighbour interpolations

robbietuk committed:
Initial commit of erfMapping

robbietuk committed:
Buildable version

robbietuk commented:
At current I am getting the following forward projection times on a very simple TOF DMI3ring scanner (29 TOF dims): [t2.txt](https://github.com/robbietuk/STIR/files/7986986/t2.txt)

This PR: 59.34 seconds
tof_ucl_sino: 78.73 seconds

KrisThielemans commented on src/include/stir/numerics/erfMapping.h:
```diff
@@ -0,0 +1,102 @@
+  std::vector<MappingTable> map;
+
+public:
+  erfMapping ()       //constructor 1 with no arguments
+  {
+    this->num_entries = 1;
+    this->lower_range = 0;
+    this->upper_range = 5;
+  }
+  erfMapping(int n)    //constructor 2 with one argument
+  {
+    this->num_entries = n;
+    this->lower_range = 0;
+    this->upper_range = 5;
+  }
```
probably these make no sense, but if they do, you should have just one with default args. Also add `explicit` such that the 1-arg constructor isn't used for conversion.

KrisThielemans commented on src/include/stir/numerics/erfMapping.h:
```diff
@@ -0,0 +1,102 @@
+    double y;
+  };
+
+  double interpolate( const std::vector<MappingTable> &data, double MappingTable::*x, double MappingTable::*y, double xValue ) const
```
not sure what these all are, but no pointers please.

KrisThielemans commented on src/include/stir/numerics/erf.inl:
```diff
@@ -419,6 +419,9 @@ erfc(double x)
 }
 
 
+
+
+
```
no spurious white spaces

KrisThielemans commented on src/include/stir/numerics/erfMapping.h:
```diff
@@ -0,0 +1,102 @@
+    this->upper_range = u;
+  }
+
+  inline int get_num_entires() const;
+  inline void set_num_entires(const int n);
```
misspelling

KrisThielemans commented on src/recon_buildblock/ProjMatrixByBin.cxx:
```diff
@@ -162,6 +162,12 @@ set_up(
     }
+
+  // Setup the custom erf code
+  my_erf.set_range(-5.0, 5.0);
```
where does the range come from?

KrisThielemans reviewed (changes requested):
WIP of course, but some comments while I looked at it.

I think `erfMapping` does not really depend on `erf` at all (i.e. the function could be passed as an argument) and is just a linear interpolator (with boundary handling).  It might be slow due to the `lower_bound` call.

If so, I'd consider using `BSplineRegularGrid` (or `BSpline1DRegularGrid` when working with `std::vector`) for all of the interpolation work. You can then choose between nearest, linear, b-spline etc, where I feel "nearest" might be sufficient for us, as long as we sample enough.
As that class assumes regular sampling, it doesn't need the `lower_bound`.

Sadly, we don't have an "x" scale in the BSpline code (i.e. the `x` is assumed to run from `samples.get_min_index()` to `samples.get_max_index()`),, so we'd need to have a `scale` and `offset` handling somewhere. 

For now, I suggest to replace the `interpolate` with a call to a `BSpline1DRegularGrid::operator((x-lower_bound)/(upper_bound-lower_bound)*num_entries)` (where you'd compute the scale only once of course).

then we see if it makes any difference at all in run-time.

robbietuk commented:
> WIP of course, but some comments while I looked at it.

Yeah, a WIP but thank you for looking. I really wanted to benchmark to ensure it was worth trying, which it looks it is.

______
I am trying to add BSplines1DRegularGrid to but keep getting an error. I have 
```
#include "stir/numerics/BSplines1DRegularGrid.h"
**********************
inline double
erfMapping::get_erf(double xp) const
{
  std::vector<double> linear_input;
  for (double i=0, imax=10; i<imax ; ++i)
    linear_input.push_back(i);
  BSpline::BSplines1DRegularGrid<double, double> spline(linear_input.begin(), linear_input.end(), BSpline::near_n);
  return spline.BSplines(xp);
}
```
Error:

```
====================[ Build | test_erf | Debug ]================================
/Applications/CLion.app/Contents/bin/cmake/mac/bin/cmake --build /Users/roberttwyman/bin/STIR/TOF/build --target test_erf -- -j 6
Scanning dependencies of target test_erf
Building CXX object src/test/numerics/CMakeFiles/test_erf.dir/test_erf.cxx.o
Linking CXX executable test_erf
Undefined symbols for architecture x86_64:
  "stir::error(char const*, ...)", referenced from:
      double stir::BSpline::BSplines_1st_der_weight<double>(double, stir::BSpline::BSplineType) in test_erf.cxx.o
      double stir::BSpline::BSplines_weights<double>(double, stir::BSpline::BSplineType) in test_erf.cxx.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [src/test/numerics/test_erf] Error 1
make[2]: *** [src/test/numerics/CMakeFiles/test_erf.dir/all] Error 2
make[1]: *** [src/test/numerics/CMakeFiles/test_erf.dir/rule] Error 2
make: *** [test_erf] Error 2

```


The error appears to be related to the `spline.BSplines(xp);` call. The scale of xp is wrong but it still fails for `spline.BSplines(0);` and `spline.BSplines(1);` etc.

robbietuk committed:
Disable interpolate, remove lower range, clean-up setup

KrisThielemans commented:
> The error appears to be related to the `spline.BSplines(xp);` call. The scale of xp is wrong but it still fails for `spline.BSplines(0);` and `spline.BSplines(1);` etc.

what do you mean? You show a linking problem. That means you shouldn't have been able to run anything.

KrisThielemans commented:
https://github.com/robbietuk/STIR/blob/7c97f1b1cf2f1088dcb8137122170c9ecc06eea2/src/test/numerics/CMakeLists.txt#L21
says that `test_erf.cxx` doesn't depend on STIR libraries, which it didn't, but it does now.

robbietuk committed:
Prevent recompilation

robbietuk committed:
erfMapping implemented

robbietuk committed:
erfMapping Test

KrisThielemans commented on src/include/stir/numerics/BSplinesDetail.inl:
```diff
@@ -17,6 +17,8 @@
 */
 #include "stir/assign.h"
 
+#ifndef BSPLINESDETAIL_H
+#define BSPLINESDETAIL_H
```
should never be necessary in a .inl file, but only in the .h

robbietuk replied:
There isnt a `BSplinesDetails.h` and I was running into issues with recompiling it more than once.

KrisThielemans replied:
ah. well, that file should have been called .h then...  sorry. In that case, call the variable `__stir_numerics_BSplinesDetail_inl_` then to avoid (unlikely) conflicts

robbietuk replied:
It appears it is used twice:

```
$grep -rn "BSplinesDetail.inl" .
Binary file ./.git/index matches
./src/include/stir/numerics/BSplinesRegularGrid.inl:19:#include "stir/numerics/BSplinesDetail.inl"
./src/include/stir/numerics/BSplines1DRegularGrid.inl:19:#include "stir/numerics/BSplinesDetail.inl"
```

KrisThielemans replied:
yes, and it breaks my rule that a .inl should not be included anywhere else than in its .h file. So you could rename the file, or just stick in the #ifdef. Probably best to do the latter, as the TOF PR is already big enough...

KrisThielemans commented on src/include/stir/numerics/erfMapping.h:
```diff
@@ -1,100 +1,81 @@
-//#include "stir/modulo.h"
-//#include "stir/common.h"
-//#include "stir/numerics/erf.h"
+#include "BSplines1DRegularGrid.h"
```
that will need full path

KrisThielemans commented on src/include/stir/numerics/erfMapping.inl:
```diff
@@ -1,62 +1,74 @@
 
 
 #include "erfMapping.h"
```
never include its own .h in an .inl (it has to be the other way around)

robbietuk committed:
Fix #include issues

robbietuk committed:
Remove spurious whitespace

robbietuk commented:
On rerunning the same forward projection test:
- This PR: 70.48 seconds
- tof_ucl_sino: 69.38 seconds

So now we are slower with BSplines (with linear interpolation) than we were with the interpolation before. It looks like the extra computation exists in `BSplies1DregularGrid::compute_BSplines_value`

robbietuk committed:
Number of samples increased

This hardly increases the computational time of BSplines

KrisThielemans commented on src/include/stir/numerics/erfMapping.inl:
```diff
@@ -0,0 +1,73 @@
+
+  //Compute a vector of erf values
+  for (int i=0; i<this->get_num_samples() ; ++i)
+    erf_values.push_back(erf(i * this->_sampling_period));
```
why only store positive values? If you want to save time, an offset is probably better than branching with `if` below (probably very little difference though)

KrisThielemans commented:
> So now we are slower with BSplines (with linear interpolation) than we were with the interpolation before. It looks like the extra computation exists in `BSplies1DregularGrid::compute_BSplines_value`

Unless there's a mistake in the code, the BSpline initialisation should happen only once. Therefore, in practice its time should be irrelevant. (It's of course surprising that it takes so long, but let's worry about that another day). I think you need to go to a larger system to see if a look-up table makes any difference.

Also, if you have enough samples, nearest-neighbour should be faster (but I don't know how much)

robbietuk committed:
Remove vector push_back

robbietuk committed:
Add get_erf_linear and get_erf_nn

robbietuk committed:
Add linear and NN tests

robbietuk committed:
erfMapping uses std::clamp and x range from -maximum_sample_value to maximum_sample_value

robbietuk committed:
Rename my_erf to erf_interpolation and use get_erf_linear

robbietuk committed:
more samples and lower higher tolerance

robbietuk committed:
refactor erfMapping method names

robbietuk committed:
Comments, documentation and simplification.

robbietuk committed:
Merge branch 'TOF/erfMappings/CustomInterpolation' into TOF/erfMapping

robbietuk commented:
@ALEXJAZZ008008 and I worked on the latest changes together this morning. We implemented a different interpolation method based upon indicies, for linear and nearest neighbour, that seems to work well.

Timings: I moved into samples (1000 samples per second) and only investigated the speed of the `apply_tof_kernel_and_symm_transformation`. I was replacing the `erf()` calls in the `get_tof_value()` method in with the new interpolation methods. The `forward_projection`s took:

- `erf(xp)`: 75.5k samples
- `get_erf_BSplines_interpolation(xp)` (linear): 63.7k samples (18.5% faster)
- `get_erf_linear_interpolation(xp)`: 49.0k samples (54.0% faster)
- `get_erf_nearest_neighbour_interpolation(xp)`: 47.3k samples (59.6% faster)

As such, I propose we use the `get_erf_linear_interpolation()` for the added accuracy and similar compuation time to `get_erf_nearest_neighbour_interpolation()`. This has essentially eliminated the cost of computing `get_tof_value()` in TOF forward projection.

___

We also removed the `erf(x) = - erf(-x)` symetry and added `std::clamp()` calls to threshold the values of `xp`.

robbietuk commented:
@KrisThielemans Thoughts?

KrisThielemans commented on src/include/stir/numerics/erfMapping.h:
```diff
@@ -0,0 +1,101 @@
+  inline void set_maximum_sample_value(double maximum_sample_value);
+
+  /*! \brief Computes the erf() values, sets up BSplines and sets up interpolation vectors.  */
+  inline void setup();
```
generally called `set_up()`

KrisThielemans commented on src/include/stir/numerics/erfMapping.inl:
```diff
@@ -0,0 +1,106 @@
+  int lower = static_cast<int>(floor(xp_in_index));
+
+  // Linear interpolation of xp between vec[lower] and vec[lower + 1]
+  return erf_values_vec[lower] + (xp_in_index - lower) * (erf_values_vec[lower + 1] - erf_values_vec[lower]);
```
are you sure you cannot get an out-of-range?

robbietuk replied:
The `test_erf.cxx` does investigate `xp` values either side of the `(-maximum_sample_value maximum_sample_value)` range for all interpolation methods

KrisThielemans replied:
yeah, either side, which the `std::clamp` will take care of. Maybe it's ok, but it seems to me you have to be careful to compute 1 entry more than you think, unless you `clamp` slightly "inside".

robbietuk replied:
Yes. We found that we needed two additional samples for the full range. I have now added a test that investigates values just inside and outside of the range. The test passes so I am fairly sure that this is correct.

KrisThielemans commented on src/include/stir/numerics/erfMapping.inl:
```diff
@@ -0,0 +1,106 @@
+    return -1.0;
+#endif
+  // Selects index of the nearest neighbour via rounding
+    return erf_values_vec[static_cast<int>((xp + this->maximum_sample_value) / this->_sampling_period)];
```
static_cast doesn't round.. you'd have to use `stir::round`. will also have to think about out-of-range

KrisThielemans commented on src/include/stir/numerics/erfMapping.h:
```diff
@@ -0,0 +1,101 @@
+
+#include "stir/numerics/BSplines1DRegularGrid.h"
+
+#ifndef ERFMAPPING_H
```
use full names to avoid coonflicts `__stir_path_filename__H__`

KrisThielemans commented on src/include/stir/numerics/erfMapping.inl:
```diff
@@ -0,0 +1,106 @@
+
+
+#include "stir/numerics/BSplines1DRegularGrid.h"
```
have to include `erf.h`

KrisThielemans reviewed (commented):
I think `erfMapping` is the wrong name. Maybe just `FastErf`?

once done, it'd be nicer to implement the interpolator as `operator()`. then we can just call `fastErf(3)`

KrisThielemans commented:
Regarding timings, it'd be nice to check why the BSpline code is slower, as it should be doing exactly the same as what you did, but I guess nobody has any appetite for that. It's not so nice to have multiple interpolators. (but stuff like debugging/fixing boundary conditions is always more work than you hope)

I'd also have preferred to make the class independent of `erf`. There really isn't anything here that depends on it. It could nicely be done via a template (in terms of a function) without any impact on speed.

but life is short...

robbietuk commented:
>I think erfMapping is the wrong name. Maybe just FastErf?

I lilke that, or `ErfInterpolation`?

robbietuk committed:
Remane and refactor class and methods

robbietuk committed:
FastErf include and define changes

robbietuk committed:
Cleanup and documentation

KrisThielemans commented:
> I lilke that, or `ErfInterpolation`?

if we rewrite with `operator()`, then it becomes a function object, so in my opinion should be called `ErfInterpolator`.

robbietuk committed:
Add actual_test_FastErf and close to limit xp value tests

robbietuk committed:
Add () operator to FastErf

robbietuk commented:
I've added the () operator. I am happy with the `FastErf` name but I can change it if needed.

robbietuk commented:
An build erorr with `std::clamp`. It appears the the command was introduced with C++17. I will revert to using `if` statments  to compile.

robbietuk committed:
Replace std::clamp (req. c++17) with min and max statements.

robbietuk commented:
Closed and moved to https://github.com/NikEfth/STIR/pull/24
